packages/gen-ai-hub/src/utils/deployment-cache.ts

```diff
@@ -0,0 +1,81 @@
+import { Cache } from '@sap-cloud-sdk/connectivity/internal.js';
+import { type AiDeployment } from '@sap-ai-sdk/ai-core';
+import { type DeploymentResolutionOptions } from './deployment-resolver.js';
+import { extractModel, type FoundationModel } from './model.js';
+
+function getCacheKey({
+  scenarioId,
+  executableId = '',
+  model,
+  resourceGroup = 'default'
+}: DeploymentResolutionOptions) {
+  return `${scenarioId}-${executableId}-${model?.name ?? ''}-${model?.version ?? ''}-${resourceGroup}`;
+}
+
+interface Deployment {
+  id: string;
+  model?: FoundationModel;
+}
+
+/**
+ * Create a cache for deployments.
+ * @param cache - Pure cache object.
+ * @returns The deployment cache.
+ * @internal
+ */
+function createDeploymentCache(cache: Cache<Deployment>) {
+  return {
+    /**
+     * Get a deployment from the cache.
+     * @param opts - Deployment resolution options to get the cached deployment for.
+     * @returns The cached deployment or undefined if not found.
+     */
+    get: (opts: DeploymentResolutionOptions): Deployment | undefined =>
+      cache.get(getCacheKey(opts)),
+    /**
+     * Store a deployment in the cache.
+     * @param opts - Deployment resolution options to set the deployment for.
+     * @param deployment - Deployment to cache.
+     */
+    set: (
+      opts: DeploymentResolutionOptions,
+      deployment: AiDeployment
+    ): void => {
+      cache.set(getCacheKey(opts), {
+        entry: transformDeploymentForCache(deployment)
+      });
+    },
+    /**
+     * Store multiple deployments in the cache, based on the model from the respective AI deployments.
+     * @param opts - Deployment resolution options to set the deployments for. Model information in the deployment resolution options are ignored.
+     * @param deployments - Deployments to retrieve the IDs and models from.
+     */
+    setAll: (
+      opts: Omit<DeploymentResolutionOptions, 'model'>,
+      deployments: AiDeployment[]
+    ): void => {
+      // go backwards to cache the first deployment ID for each model
+      [...deployments].reverse().forEach(deployment => {
+        cache.set(getCacheKey({ ...opts, model: extractModel(deployment) }), {
```

Comment:
I think this strategy of building the cache-key is a bit awkward and for the main use case probably even broken?

Broken: The model version is part of the cache key, but when users query using only the model name, a cache miss apparently occurs. See [this comment](https://github.com/SAP/ai-sdk-js/pull/96/files#r1741103818). Or am I missing something?

Awkward: We are partially building the cache key from the actual result, and partially from the user input. I guess the issue here is that `executableId` is not part of the response payload. This again leads to unnecessary cache misses where the same deployment is queried, sometimes with and other times without the executable ID. But I guess since it's not part of the response we don't have much of a choice here. We could query configurations in addition (which have the executable ID), but that would probably be over-optimizing 😄

Reply:
> I think this strategy of building the cache-key is a bit awkward and for the main use case probably even broken?
> 
> Broken: The model version is part of the cache key, but when users query using only the model name, a cache miss apparently occurs. See [this comment](https://github.com/SAP/ai-sdk-js/pull/96/files#r1741103818). Or am I missing something?

Good catch, I will add this in a follow up.
 
> Awkward: We are partially building the cache key from the actual result, and partially from the user input. I guess the issue here is that `executableId` is not part of the response payload. This again leads to unnecessary cache misses where the same deployment is queried, sometimes with and other times without the executable ID. But I guess since it's not part of the response we don't have much of a choice here. We could query configurations in addition (which have the executable ID), but that would probably be over-optimizing 😄

I thought about that as well, but at least when you use the same options twice you get a cached result, so might be ok. The other problem is worse ;)